### PR TITLE
[ADF-3401] Search filters - fix facet update

### PR DESCRIPTION
**Pull request by suzanadirla**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [ ] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [x] Bugfix
> - [ ] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)
https://issues.alfresco.com/jira/browse/ADF-3401
The new facet items received in the response are not added to the search filter component.

**What is the new behaviour?**
When a facet is received in the response and is already displayed, then only its bucket count is updated, otherwise, **the new facet is now added to the displayed items of the search filter component**. 
The already displayed facets are removed when they are no longer received in the response. For better UX, such ‘empty’ facets are kept on the categories while is a selection done by the user. 
Any filtering is applied also to the new facets received from response and any facet selections done by the user will not be lost. 
Two clean up buttons are now displayed to make it easier for the user to remove all selections or reset entire search filters. 

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
dependent on [[ADF-3496] Grouped facet queries #4209](https://github.com/Alfresco/alfresco-ng2-components/pull/4209)

### CODE REVIEW

**Comment by suzanadirla** [comment]
I've just checked the Jira bug's description and video and indeed those are not so relevant. I will update the bug with a more relevant video and description. Sorry about this, Eugenio.

File: lib/content-services/search/components/search-filter/search-filter.component.ts [hunk 1]
```diff
@@ -160,59 +189,76 @@ export class SearchFilterComponent implements OnInit, OnDestroy {
...
             const responseField = (context.facets || []).find((response) => response.type === 'field' && response.label === field.label) || {};
             const responseBuckets = this.getResponseBuckets(responseField);
 
-            const bucketList = new SearchFilterList<FacetFieldBucket>(responseBuckets, field.pageSize);
-            bucketList.filter = (bucket: FacetFieldBucket): boolean => {
-                if (bucket && bucketList.filterText) {
-                    const pattern = (bucketList.filterText || '').toLowerCase();
-                    const label = (this.translationService.instant(bucket.display) || this.translationService.instant(bucket.label)).toLowerCase();
-                    return this.queryBuilder.config.filterWithContains ? label.indexOf(pattern) !== -1 : label.startsWith(pattern);
+            if (alreadyExistingField) {
```
[the remark this answers is not in the document]
Comment (suzanadirla) [reply lib/content-services/search/components/search-filter/search-filter.component.ts]: I have made some code refactorings and it should be better now. Please check, thank you!

File: lib/content-services/search/components/search-filter/models/search-filter-list.model.ts [hunk 2]
```diff
@@ -109,6 +109,18 @@ export class SearchFilterList<T> implements Iterable<T> {
         this.filterText = '';
     }
 
+    addItem(item) {
```
Comment (DenysVuika) [inline lib/content-services/search/components/search-filter/models/search-filter-list.model.ts]: missing type definition and null checks

File: lib/content-services/search/components/search-filter/models/search-filter-list.model.ts [hunk 3]
```diff
@@ -109,6 +109,18 @@ export class SearchFilterList<T> implements Iterable<T> {
...
+        this.items.push(item);
+        this.applyFilter();
+    }
+
+    deleteItem(item) {
```
Comment (DenysVuika) [inline lib/content-services/search/components/search-filter/models/search-filter-list.model.ts]: missing type definition and null checks

File: lib/content-services/search/components/search-filter/models/search-filter-list.model.ts (hunk 2 above)
Comment (suzanadirla) [reply lib/content-services/search/components/search-filter/models/search-filter-list.model.ts]: I made the changes, thanks Denys!

File: lib/content-services/search/components/search-filter/search-filter.component.ts [hunk 4]
```diff
@@ -342,4 +356,55 @@ export class SearchFilterComponent implements OnInit, OnDestroy {
...
+    private getApiResponseField(context: ResultSetContext, itemType: string, fieldLabel: string): GenericFacetResponse {
+        return (context.facets || []).find((response) => response.type === itemType && response.label === fieldLabel) || {};
+    }
+
+    private getAlreadyDisplayedField(itemType: string, fieldLabel: string): FacetField {
+        return (this.responseFacets || []).find((response) => response.type === itemType && response.label === fieldLabel);
+    }
+
+    private updateExistingBuckets(responseField, responseBuckets, alreadyExistingField, alreadyExistingBuckets) {
+        const shouldDelete = [];
```
Comment (DenysVuika) [inline lib/content-services/search/components/search-filter/search-filter.component.ts]: Rename the variable to express intent better
```suggestion
        const bucketsToDelete = [];
```

File: lib/content-services/search/components/search-filter/search-filter.component.ts [hunk 5]
```diff
@@ -342,4 +356,55 @@ export class SearchFilterComponent implements OnInit, OnDestroy {
 
         return `${fieldName}:${startLimit}"${start}" TO "${end}"${endLimit}`;
     }
+
+    private getApiResponseField(context: ResultSetContext, itemType: string, fieldLabel: string): GenericFacetResponse {
```
Comment (DenysVuika) [inline lib/content-services/search/components/search-filter/search-filter.component.ts]: Rename the function as all functions are APIs
```suggestion
    private findFacet(context: ResultSetContext, itemType: string, fieldLabel: string): GenericFacetResponse {
```

File: lib/content-services/search/components/search-filter/search-filter.component.ts [hunk 6]
```diff
@@ -342,4 +356,55 @@ export class SearchFilterComponent implements OnInit, OnDestroy {
...
+        return (context.facets || []).find((response) => response.type === itemType && response.label === fieldLabel) || {};
+    }
+
+    private getAlreadyDisplayedField(itemType: string, fieldLabel: string): FacetField {
```
Comment (DenysVuika) [inline lib/content-services/search/components/search-filter/search-filter.component.ts]: Rename to better show what it does
```suggestion
    private findResponseFacet(itemType: string, fieldLabel: string): FacetField {
```

Comment (suzanadirla) [reply lib/content-services/search/components/search-filter/search-filter.component.ts]: I applied the suggested renamings. Thank you, @DenysVuika

**Pull request merged**